### PR TITLE
add shell script-driven testing

### PR DESCRIPTION
**Pull request by mccanne**
This commit provides a framework for tests run
as shell commands.  A test is defined with
input and expected output files the actual files
are checked against the expected file output.

An example (and useful test) is provided in
package tests/suite/diropt, which tests the
functionality of the -d and -o flags.

### CODE REVIEW

File: pkg/test/script.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	src += "cd " + s.dir + "\n"
+	src += s.Script
+	_, err = f.Write([]byte(src))
+	if err != nil {
+		f.Close()
+		return "", "", err
+	}
+	f.Close()
+	s.createInputFiles()
+	cmd := exec.Command("/bin/bash", scriptName)
```
Comment (mattnibs) [inline pkg/test/script.go]:
```suggestion
	cmd := exec.Command("/bin/bash", scriptName)
	cwd, _ := os.Getwd()
	cmd.Env = []string{"PATH="+filepath.Join(cwd, "dist")}
```

File: pkg/test/exec.go [hunk 2]
```diff
@@ -13,8 +13,12 @@ type Exec struct {
 	Expected string
 }
 
-func (e *Exec) Run() (string, error) {
-	cmd := exec.Command("/bin/bash", "-c", "PATH=$PATH:$GOPATH/bin ; "+e.Command)
+func (e *Exec) Run(path string) (string, error) {
+	src := e.Command
+	if path != "" {
+		src = "PATH:$PATH:" + path + " ; " + src
```
Comment (mattnibs) [inline pkg/test/exec.go]: If someone has zq in path already will this hit the right zq?

File: pkg/test/script.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	cwd, _ := os.Getwd()
+	cmd.Env = []string{"PATH=" + filepath.Join(cwd, "dist")}
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err = cmd.Run()
+	return string(stdout.Bytes()), string(stderr.Bytes()), err
```
Comment (mattnibs) [inline pkg/test/script.go]:
```suggestion
	return stdout.String(), stderr.String(), err
```

File: tests/queries_test.go [hunk 4]
```diff
@@ -1,15 +1,23 @@
-// +build system
-
```
Comment (mattnibs) [inline tests/queries_test.go]: One man's opinion but I like this instead of the test short stuff

**Comment by mattnibs** [comment]
You should add `tests/test-root` to `.gitignore`

File: tests/queries_test.go [hunk 5]
```diff
@@ -21,11 +29,54 @@ func TestInternal(t *testing.T) {
...
+				if !assert.Exactly(t, file.Data, actual, "Wrong shell script results") {
+					fail = true
+				}
+			}
+			if !fail {
+				// Remove the testdir on success.  If test fails,  we
+				// leave it behind in testroot for debugging.  These
+				// failed test directories have to be manually removed.
+				shell.Cleanup()
+			}
```
Comment (mattnibs) [inline tests/queries_test.go]: Feels like root-dir should be cleared on the next run... my started getting a lot of gunk in it.

File: pkg/test/script.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+func NewShellTest(s Shell) *ShellTest {
+	return &ShellTest{Shell: s}
+}
+
+// Setup creates a temp directory in the dir path provided.
+func (s *ShellTest) Setup(dir string) (*os.File, error) {
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+	d, err := ioutil.TempDir(dir, "test")
```
Comment (mattnibs) [inline pkg/test/script.go]: Can we use the name of the test? All these random dirs made it hard to know what files belonged to what.

File: tests/queries_test.go (hunk 4 above)
Comment (mccanne) [reply tests/queries_test.go]: Yes, it's better.  I was having random trouble and commented this out so I'll put it back.

File: pkg/test/script.go (hunk 3 above)
Comment (mccanne) [reply pkg/test/script.go]: +1

File: tests/queries_test.go (hunk 5 above)
Comment (mccanne) [reply tests/queries_test.go]: I kind of like the idea but it doesn't feel like expected behavior and could cause surprises.

File: pkg/test/script.go (hunk 6 above)
Comment (mccanne) [reply pkg/test/script.go]: yes, I meant to do that.

Comment (mccanne) [reply pkg/test/script.go]: Actually, I meant to use the name in a tempdir, but I like your idea better.  This is my preference to auto-delete...  we clear out the old test and replace it with the new one to keep the clutter down.  It does mean we need to be careful to not give the same test the same name (which I have done when you cut and paste) then run them in parallel.  I'll put a check in for that.

**Comment by mccanne** [comment]
> You should add `tests/test-root` to `.gitignore`

done

**Pull request merged**